### ISSUE TITLE
predict_generator increments batch_index by 1 too many on each call

### ISSUE THREAD

**Issue by mcamack** [keras-team/keras#8002]
I'm using latest keras 2.0.8 with python 3.5.2 and tensorflow

Calling predict_generator, the batch_index seems to skip ahead by 1. I must not understand how this works. It seems to skip ahead by max_queue_size, is it actually reading those examples at some point? In the example below, I get 10 batches, but then my index is at 20. If I change queue to 1, it would be at 11. I tried keras 1.2.2 and queue size didn't have any effect, but it would always skip ahead by 1 extra batch index. 

`batch_size = 32`
`IDG_sub = image.ImageDataGenerator()`
`gen_submission_test_batches = IDG_sub.flow_from_directory("file/path", target_size=(224,224), color_mode="rgb", class_mode=None, batch_size=batch_size, shuffle=False)`
`gen_submission_test_batches.batch_index=0`
`gen_submission_test_batches.batch_index`
0

`submission_preds = model.predict_generator(gen_submission_test_batches, steps=10, max_queue_size=10, workers=1, use_multiprocessing=False, verbose=1)`
10/10 [==============================] - 2s 

`gen_submission_test_batches.batch_index`
20

Trying to iterate using next() it only advances the batch index by 1:
`gen_submission_test_batches.batch_index=0`
`gen_submission_test_batches.batch_index`
0

`tmp=next(gen_submission_test_batches)`
`gen_submission_test_batches.batch_index`
1

I am predicting on a lot of data with thousands of classes, so I'm doing it in chunks. When it gets to the final chunk I get the "could not broadcast input array from shape" error. Checking total_batches_seen at that point puts the number over what the total should be for the whole dataset in 1 epoch. I'm just confused about how the "step" and "max_queue_size" arguments effect predict_generator. Am I supposed to roll the batch_index back by max_queue_size after every call to predict_generator?

**Comment by nikoladze** [comment]
I think the reason for this is that by default the batches are put into a queue by a separate thread and the main generator just fetches the batches from that queue. If you don't want need use multiple workers this behaviour can be avoided by passing `workers=0` in `predict_generator`.

**Comment by mclipsgh** [comment]
I did experience the same problem and think this needs to be resolved. The workers=0 solution works, but the default (workers=1 and max_queue_size=10) would cause the output to delay for 10 batches and cause a lot of head-scratchers.
